dbr quant: add limited support for `torch.nn.ModuleList`

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#70372 dbr quant: add limited support for `torch.nn.ModuleList`**
* #70349 dbr quant support for custom leaf modules, part 3/x
* #70335 dbr quant: support for custom leaf modules, part 2/x
* #70330 dbr quant: support for custom leaf modules, part 1/x
* #70329 dbr quant: unbreak case when child module not returning any outputs
* #70328 dbr quant: support functional linear without bias
* #70257 dbr quant: support dynamic linear
* #70256 dbr quant: do not attach auto_quant_state to observers

Summary:

Enables basic support for `torch.nn.ModuleList` in DBR quant
by stopping it from being a leaf.  For now, we
require the user to check for `AutoQuantizationState` if they are
looping over the contents without any bounds checking.

In future PRs, we can explore how to solve this without requiring
user code changes.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_module_list
```

Differential Revision: [D33302329](https://our.internmc.facebook.com/intern/diff/D33302329)

vkuzo committed:
dbr quant: add limited support for `torch.nn.ModuleList`

Summary:

Enables basic support for `torch.nn.ModuleList` in DBR quant
by stopping it from being a leaf.  For now, we
require the user to check for `AutoQuantizationState` if they are
looping over the contents without any bounds checking.

In future PRs, we can explore how to solve this without requiring
user code changes.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_module_list
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33302329).

vkuzo committed:
Update on "dbr quant: add limited support for `torch.nn.ModuleList`"


Summary:

Enables basic support for `torch.nn.ModuleList` in DBR quant
by stopping it from being a leaf.  For now, we
require the user to check for `AutoQuantizationState` if they are
looping over the contents without any bounds checking.

In future PRs, we can explore how to solve this without requiring
user code changes.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_module_list
```

Differential Revision: [D33302329](https://our.internmc.facebook.com/intern/diff/D33302329)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33302329).

jerryzh168 commented on test/quantization/dbr/test_quantize_dbr.py:
```diff
@@ -704,6 +705,33 @@ def forward(self, x):
         FileCheck().check_count("aten::add", 1, exactly=True).run(mqt.m2.graph)
         FileCheck().check_count("quantized::add", 0, exactly=True).run(mqt.m2.graph)
 
+    def test_module_list(self):
```
nit: just to confirm, do we have a test for Sequential as well? also for completeness do we want to add support for ModuleDict?

vkuzo replied:
yes, sequential is already tested.  ModuleDict could be added in a future PR, it's just not needed for the current model we are working on.

vkuzo committed:
Update on "dbr quant: add limited support for `torch.nn.ModuleList`"


Summary:

Enables basic support for `torch.nn.ModuleList` in DBR quant
by stopping it from being a leaf.  For now, we
require the user to check for `AutoQuantizationState` if they are
looping over the contents without any bounds checking.

In future PRs, we can explore how to solve this without requiring
user code changes.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_module_list
```

Differential Revision: [D33302329](https://our.internmc.facebook.com/intern/diff/D33302329)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33302329).
